machineconfiguration/v1/types.go

```diff
@@ -0,0 +1,654 @@
+package v1
+
+import (
+	configv1 "github.com/openshift/api/config/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// MachineConfigRoleLabelKey is metadata key in the MachineConfig. Specifies the node role that config should be applied to.
+// For example: `master` or `worker`
+const MachineConfigRoleLabelKey = "machineconfiguration.openshift.io/role"
+
+// KubeletConfigRoleLabelPrefix is the label that must be present in the KubeletConfig CR
+const KubeletConfigRoleLabelPrefix = "pools.operator.machineconfiguration.openshift.io/"
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ControllerConfig describes configuration for MachineConfigController.
+// This is currently only used to drive the MachineConfig objects generated by the TemplateController.
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+// +openshift:compatibility-gen:internal
+type ControllerConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// TODO(jkyros): inconsistent historical generation resulted in the controllerconfig CRD being
+	// generated with all fields required, while everything else was generated with optional
+
+	// +kubebuilder:validation:Required
+	// +required
+	Spec ControllerConfigSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
```

Comment:
Hmm that failed integration test made me poke through all the required fields. 

The MCO hasn't been historically great at required/non-required fields since most of our CRD updates had been done manually.  It looks like ControllerConfig was originally (years ago) generated with "required by default" at the package level and has been updated that way, while everything else was generated with "optional by default" at the package level and maintained that way.  

This PR initially defaulted to "fields required by default" at the package level. Looking at the rest of the APIs in the repo, it looks like there's a mix?

I've adjusted this now so that:
-  `// +kubebuilder:validation:Optional` is set at the package level 
- the `ControllerConfig` fields that were previously required are set explicitly to required 
- the tests should be fixed and now supply the required fields (I will clean the data up before merge) 

Generally though, what are your "best practice" thoughts on how required fields should be handled: 
- Required by default, mark everything we don't want required as optional or 
- Optional by default, mark everything we don't want optional as required 
- Mark literally every field explicitly so it doesn't matter? :smile: